templated-this.md

```diff
@@ -1,151 +1,315 @@
+  bar([s] (auto&& this) { 
+    return std::forward<decltype(this)>(this).s;
+  }
+}
+```
+
+What does `this` in a parameter list mean?
+------------------------------------------
+
+The meaning of the different ways to pass `this` is very similar to the way
+`catch` works. This table uses `T` to refer to the enclosing class. `T` is not a
+`template`. This shall be illustrated on the method family `void f(...)...`.
+
+| written as                        | C++17 signature             | comments   | 
+| -------------------------         | --------------------------- | ---------- | 
+| `void f(T this)`                  | currently not available     | [value]    | 
+| `void f(T& this)`                 | `void f()&`                 |            | 
+| `void f(T&& this)`                | `void f()&&`                |            | 
+| `void f(T const this)`            | currently not available     | [value]    | 
+| `void f(T const& this)`           | `void f() const&`           |            | 
+| `void f(T const&& this)`          | `void f() const&&`          |            | 
+| `void f(T volatile this)`         | currently not available     | [value]    | 
+| `void f(T volatile& this)`        | `void f() volatile&`        |            | 
+| `void f(T volatile&& this)`       | `void f() volatile&&`       |            | 
+| `void f(T const volatile this)`   | currently not available     | [value]    | 
+| `void f(T const volatile& this)`  | `void f() const volatile&`  |            | 
+| `void f(T const volatile&& this)` | `void f() const volatile&&` |            | 
+
+*Notes:*
+
+- *[value]*: whether passing by value should be allowed is debatable, but seems
+  obviously desired for completeness and parity with inline friend functions.
+- The interpretation of `this` in the method body differs, but only one
+  definition for a given signature may be present, eg. one may define at most
+  one of `void f()&`, or `void f(T& this)` or `void f()`, the first and last
+  already being exclusive of one another.
+
+
+### How does templated `this` work?
+
+Using existing deduction rules for template parameters, which will deduce the
+type of `this` to something in the above table.
+
+
+### What does `this` mean in the body of a method?
+
+It behaves exactly as a regular parameter declared in the same way.
+
+
+### Constructors
+
+No exceptions to above rules. If a particular constructor signature is not
+allowed by the language, it continues to be disallowed. We can already access
+already-initialized members in initialization lists, which means `this` is
+already available, even though it hasn't been completely constructed yet.
+
+
+### What about pass by value methods?
+
+We think they are a logical extension of the mechanism, and would go a long way
+towards making methods as powerful as inline friend functions, with the only
+difference being the call syntax.
+
+One really cool implication of this is that the `this` parameter would be
```

Comment:
Strike 'really cool'. It is really cool, but let's be impartial :)

Reply:
heh, yeah... :(

Reply:
People more often think things are cool if they come to the conclusion rather than being told things are cool, that's all :)